### PR TITLE
Add constraints for values added by DeclarationExpression

### LINKED ISSUE #2936: Extend CSharpExplodedGraph with support for DeclarationExpressionSyntax
### Description

Currently the `CSharpExplodedGraph` is ignoring `DeclarationExpressionSyntax` nodes and there are no symbolic values created for the declared variables.

For example, in case of: `dictionary.TryGetValue(key, out var value);`, the variable **value** should be handled and added to the list of values.

Additionally we should add constraints for the new symbolic value.

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes #2936

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 1]
```diff
@@ -495,7 +495,7 @@ public void ExplodedGraph_StaticLocalFunctions()
 
         [TestMethod]
         [TestCategory("Symbolic execution")]
-        public void ExplodedGraph_DeclarationStatementVisit()
+        public void ExplodedGraph_DeclarationExpressionVisit_AsOutParameter()
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: I think the name should mention that this is testing a value type declaration

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 2]
```diff
@@ -504,16 +504,15 @@ namespace Namespace
 {
   public class DeclarationStatement
   {
-    public int Main(Dictionary<string, string> dictionary, string key)
+    public int Main(Dictionary<string, int> dictionary, string key)
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: do we still have a UT for reference type declarations? because I see here only value-type ones

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 3]
```diff
@@ -538,22 +537,58 @@ public int Main(Dictionary<string, string> dictionary, string key)
...
             context.WalkWithInstructions(5);
         }
 
+        [TestMethod]
+        [TestCategory("Symbolic execution")]
+        public void ExplodedGraph_DeclarationExpressionVisit_AsOutParameter_Discard()
+        {
+            const string testInput = "string key = null; Dictionary<string, int> dictionary = new Dictionary<string, int>(); dictionary.TryGetValue(key, out var _); dictionary.TryGetValue(key, out _);";
+            var context = new ExplodedGraphContext(testInput);
+            var dictionarySymbol = context.SemanticModel.GetDeclaredSymbol(context.MainMethod.ParameterList.Parameters.First());
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: The `dictionarySymbol` here is not the actual disctionary symbol. The default main has a `bool` as parameter.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 4]
```diff
@@ -538,22 +537,58 @@ public int Main(Dictionary<string, string> dictionary, string key)
...
+            context.ExplodedGraph.InstructionProcessed +=
+                (sender, args) =>
+                {
+                    var instruction = args.Instruction.ToString();
+                    switch (instruction)
+                    {
+                        case "dictionary.TryGetValue(key, out var _)":
+                            args.ProgramState.GetSymbolValue(dictionarySymbol).Should().NotBeNull();
+                            dictionarySymbol.HasConstraint(ObjectConstraint.NotNull, args.ProgramState).Should().BeTrue();
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: Not sure what this is testing here as this is not related to out parameter or discard designation

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 5]
```diff
@@ -538,22 +537,58 @@ public int Main(Dictionary<string, string> dictionary, string key)
...
+
+                            args.ProgramState.HasValue.Should().BeFalse();
+                            break;
+
+                        case "dictionary.TryGetValue(key, out _)":
+                            args.ProgramState.GetSymbolValue(dictionarySymbol).Should().NotBeNull();
+                            dictionarySymbol.HasConstraint(ObjectConstraint.NotNull, args.ProgramState).Should().BeTrue();
+
+                            args.ProgramState.HasValue.Should().BeFalse();
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: Before checking that the program state has empty stack here, it would be nice to check that it has value on the stack after having evaluated the `key` parameter

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 6]
```diff
@@ -538,22 +537,58 @@ public int Main(Dictionary<string, string> dictionary, string key)
...
+                            break;
+                    }
+                };
+
+            context.WalkWithInstructions(14);
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: It has already happen in the past that some cases of the switch above are not explored: it would be nice to add a counter that increments for each of the cases, so that we know for sure all of them were explored.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs (hunk 2 above)
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: No, indeed. I've added a new test to assert that we don't add the **NotNull** constraint for reference types.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs (hunk 3 above)
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: 😄 very good catch. It was a copy paste error which I've fixed now.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs (hunk 4 above)
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: Indeed this was not meant to test the parameter but the actual dictionary constraints.
